File: hash_tables/hash_tables.py
class Element:

    __slots = ['key', 'next', 'prev']

    def __init__(self, key):
        self.key = key
        self.next = self.prev = None


class HashTable:

    __slots__ = ['bucket_size', 'tables']

    def __init__(self, bucket_size=13):
        self.bucket_size = bucket_size
        self.tables = [None for _ in range(bucket_size)]

    def put(self, z):
        hash_val = self.__calculate_hash_val(z.key)

        z.next = self.tables[hash_val]
        if self.tables[hash_val]:
            self.tables[hash_val].prev = z
        self.tables[hash_val] = z

    def contains(self, key):
        hash_val = self.__calculate_hash_val(key)

        p = self.tables[hash_val]
        while p and p.key != key:
            p = p.next

        return p

    def remove(self, z):
        hash_val = self.__calculate_hash_val(z.key)

        if z.prev:
            z.prev.next = z.next
        else:
            self.tables[hash_val] = z.next
        if z.next:
            z.next.prev = z.next

        z = None

    def print_hash_table(self):
        for i in range(self.bucket_size):
            print(f"{i} : ", end='')

            p = self.tables[i]
            while p:
                print(f"{p.key}", end=' -> ' if p.next else '')
                p = p.next

            print('')

    def __calculate_hash_val(self, key):
        return key % self.bucket_size


if __name__ == '__main__':
    ht = HashTable()

    for i in range(1, 101):
        ht.put(Element(i))

    ht.print_hash_table()

    while True:
        print("1:put 2:remove 3:print > ", end='')
        op = int(input())

        if op == 1:
            print("input key > ", end='')
            key = int(input())

            ht.put(Element(key))
        elif op == 2:
            print("input key > ", end='')
            key = int(input())

            z = ht.contains(key)
            if not z:
                print(f"{key} is not found in the hash table.")
            else:
                ht.remove(z)
        elif op == 3:
            ht.print_hash_table()
        else:
            print("invalid operation")

File: hash_tables/hash_tables.go
package main

import "fmt"

type Element struct {
	key  int
	next *Element
	prev *Element
}

type Hashtable struct {
	bucketSize int
	table      []*Element
}

func (ht *Hashtable) put(z *Element) {
	hashVal := calculateHashValue(ht, z.key)

	if ht.table[hashVal] != nil {
		ht.table[hashVal].prev = z
	}
	z.next = ht.table[hashVal]
	ht.table[hashVal] = z
}

func (ht *Hashtable) contains(key int) *Element {
	hashVal := calculateHashValue(ht, key)

	for p := ht.table[hashVal]; p != nil; p = p.next {
		if p.key == key {
			return p
		}
	}

	return nil
}

func (ht *Hashtable) remove(z *Element) {
	hashVal := calculateHashValue(ht, z.key)

	if z.next != nil {
		z.next.prev = z.prev
	}
	if z.prev != nil {
		z.prev.next = z.next
	} else {
		ht.table[hashVal] = z.next
	}
	z = nil
}

func (ht *Hashtable) printHashtable() {
	for i := 0; i < ht.bucketSize; i++ {
		fmt.Printf("%d : ", i)

		for p := ht.table[i]; p != nil; p = p.next {
			fmt.Printf("%d", p.key)
			if p.next != nil {
				fmt.Print(" -> ")
			}
		}
		fmt.Println()
	}
}

func calculateHashValue(ht *Hashtable, key int) int {
	return key % ht.bucketSize
}

func main() {
	ht := Hashtable{13, nil}
	ht.table = make([]*Element, ht.bucketSize)

	for i := 1; i <= 100; i++ {
		z := Element{i, nil, nil}
		ht.put(&z)
	}

	for {
		fmt.Print("1:put 2:remove 3:print > ")
		var op int
		fmt.Scanf("%d", &op)

		switch op {
		case 1:
			fmt.Print("input key > ")
			var data int
			fmt.Scanf("%d", &data)

			z := Element{data, nil, nil}
			ht.put(&z)
		case 2:
			fmt.Print("input data > ")
			var data int
			fmt.Scanf("%d", &data)

			z := ht.contains(data)
			if z == nil {
				fmt.Printf("%d is not found in the hash table.\n", data)
			} else {
				ht.remove(z)
			}
		case 3:
			ht.printHashtable()
		default:
			fmt.Println("invalid operation")
		}
	}
}
